packages/@apphosting/community/adapter-astro/src/index.ts

```diff
@@ -0,0 +1,115 @@
+import * as fs from "node:fs/promises";
+import { stringify as yamlStringify } from "yaml";
+import { fileURLToPath } from "url";
+import { dirname } from "path";
+import fsExtra from "fs-extra";
+import { exec } from "child_process";
+import { createRequire } from "module";
+import path from "path";
+import type { AstroAdapter, AstroIntegration } from "astro";
+import { AstroError } from "astro/errors";
+import type { Options, UserOptions } from "./types.js";
+import { OutputBundleConfig, Availability } from "@apphosting/common";
+export const { move, exists, writeFile, readJson, readdir, readFileSync, existsSync, mkdir } =
+  fsExtra;
+
+export function getAdapter(options: Options): AstroAdapter {
+  const require = createRequire(import.meta.url);
+  const serverEntrypoint = path.join(require.resolve("@astrojs/node"), "../server.js");
+  const previewEntrypoint = path.join(require.resolve("@astrojs/node"), "../preview.js");
+  return {
+    name: "@apphosting/astro-adapter",
+    serverEntrypoint: serverEntrypoint,
+    previewEntrypoint: previewEntrypoint,
+    exports: ["handler", "startServer", "options"],
+    args: options,
+    adapterFeatures: {
+      buildOutput: "server",
+      edgeMiddleware: false,
+    },
+    supportedAstroFeatures: {
+      hybridOutput: "stable",
+      staticOutput: "stable",
+      serverOutput: "stable",
+      sharpImageService: "stable",
+      i18nDomains: "experimental",
+      envGetSecret: "stable",
+    },
+  };
+}
+
+export default function createIntegration(userOptions: UserOptions): AstroIntegration {
+  if (!userOptions?.mode) {
+    throw new AstroError(`Setting the 'mode' option is required.`);
+  }
+
+  let _options: Options;
+  return {
+    name: "@apphosting/astro-adapter",
+    hooks: {
+      "astro:config:setup": ({ updateConfig, config }) => {
+        updateConfig({
+          image: {
+            endpoint: config.image.endpoint ?? "astro/assets/endpoint/node",
+          },
+          vite: {
+            ssr: {
+              noExternal: ["@apphosting/astro-adapter"],
+            },
+          },
+        });
+      },
+      "astro:config:done": ({ setAdapter, config }) => {
+        _options = {
+          ...userOptions,
+          client: config.build.client?.toString(),
+          server: config.build.server?.toString(),
+          host: config.server.host,
+          port: config.server.port,
+          assets: config.build.assets,
+        };
+        setAdapter(getAdapter(_options));
+      },
+      "astro:build:done": async () => {
+        await fs.mkdir("./.apphosting");
```

Comment:
```suggestion
        await fs.mkdir("./.apphosting", {
          // Don't throw if directory already exists. Important when debugging the build locally
          recursive: true
        });
```

https://nodejs.org/dist/latest-v16.x/docs/api/fs.html#fsmkdirpath-options-callback:~:text=.%20Calling%20fs.mkdir()%20when%20path%20is%20a%20directory%20that%20exists%20results%20in%20an%20error%20only%20when%20recursive%20is%20false